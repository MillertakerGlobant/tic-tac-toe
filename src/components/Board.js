import React, { useState } from 'react';
import BoardItem from './BoardItem';
import sadBoy from '../assets/game-over.png';
import winnerBoy from '../assets/winner.png';
import { calculateGame } from '../utils/Helper';

const Board = () => {
  const [board, setBoard] = useState([-1,-1,-1,-1,-1,-1,-1,-1,-1]);
  const [turn, setTurn] = useState(0);
  const [gameOver, setGameOver] = useState([false, null]);
  const [winner, setWinner] = useState(null);
  const player = ['x','o'];
  const [prevBoard, setPrevBoard] = useState([]);

  function clickItem(e) {
    if(!gameOver[0] && winner === null) {
      var index = e.target.getAttribute('arrayindex');
      var currentBoard = JSON.parse(JSON.stringify(board));
      setPrevBoard(JSON.parse(JSON.stringify(board)));

      turn === 8 && setGameOver([true, null])
      currentBoard[index] = turn % 2;
      
      setTurn(turn + 1);
      setBoard(currentBoard); 
      setWinner(calculateGame(currentBoard))
    }
  }

  function resetGame(e) {
    e.preventDefault();
    setBoard([-1,-1,-1,-1,-1,-1,-1,-1,-1]);
    setTurn(0);
    setWinner(null);
    setGameOver([false, null])
    setPrevBoard([]);
  }

  function rewindGame(e) {
    e.preventDefault();

    if(!gameOver[0] && winner === null) {
      setBoard(prevBoard);
      setPrevBoard([]);
      setTurn(turn - 1);
    }
  }

  const boardItems = board.map((e, i) => <BoardItem key={`boardItem${i}`} arrayindex={i} value={e} onClick={clickItem} ></BoardItem> )
  
  return (
    <div className="gameWrapper">
      Game Board 

      <ul className="BoardWrapper">
        { boardItems }
      </ul>

      <a className="resetBtn" onClick={resetGame} href="">Reset Game</a>
      { winner === null && <p>Player {player[turn % 2]} plays</p>  }
      { prevBoard.length > 0 &&  <a className="rewindBtn" onClick={rewindGame} href="">Rewind Move</a> }

      {gameOver[0] && winner === null && <div> <img className="gameOver" src={sadBoy} alt=""/> <h3>GAME OVER</h3> </div>  }
      {winner !== null && <div> <img className="gameOver" src={winnerBoy} alt=""/> <h3> { player[winner] } player is the winner</h3> </div>  }
    </div>
  )
}

export default Board;