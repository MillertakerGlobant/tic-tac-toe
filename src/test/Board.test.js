import { shallow, mount } from 'enzyme';
import Board from '../components/Board';

describe('<Board />  Component test', () => {
  var wrapper = shallow(<Board />);
  beforeEach(() => {
    wrapper = shallow(<Board />);
  });

  test('renders empty Board component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('test, turn 1 player 1', () => {
    wrapper = mount(<Board />);
    wrapper.find('ul').children().at(0).simulate('click');
    const boardItem = wrapper.find('ul').children().at(0).text().trim();
    expect(boardItem).toBe('x');
  });

  test('test, turn 1 player 2', () => {
    wrapper = mount(<Board />);
    wrapper.find('ul').children().at(0).simulate('click');
    wrapper.find('ul').children().at(2).simulate('click');
    const boardItem = wrapper.find('ul').children().at(2).text().trim();
    expect(boardItem).toBe('o');
  });

  test('test, turn 2 player 2', () => {
    wrapper = mount(<Board />);
    wrapper.find('ul').children().at(0).simulate('click');
    wrapper.find('ul').children().at(2).simulate('click');
    wrapper.find('ul').children().at(1).simulate('click');
    const boardItem = wrapper.find('ul').children().at(1).text().trim();
    expect(boardItem).toBe('x');
  });

  test('draw game board', () => {
    wrapper = mount(<Board />);
    wrapper.find('ul').children().at(0).simulate('click');
    wrapper.find('ul').children().at(1).simulate('click');
    wrapper.find('ul').children().at(2).simulate('click');
    wrapper.find('ul').children().at(4).simulate('click');
    wrapper.find('ul').children().at(7).simulate('click');  
    wrapper.find('ul').children().at(6).simulate('click');  
    wrapper.find('ul').children().at(8).simulate('click');  
    wrapper.find('ul').children().at(5).simulate('click');  
    wrapper.find('ul').children().at(3).simulate('click');  

    console.log(wrapper.debug());

    const gameOverText = wrapper.find('h3').text().trim();
    expect(gameOverText).toBe('GAME OVER');
  });

  test('Player 1 winner', () => {
    wrapper = mount(<Board />);
    wrapper.find('ul').children().at(0).simulate('click');
    wrapper.find('ul').children().at(2).simulate('click');
    wrapper.find('ul').children().at(3).simulate('click');
    wrapper.find('ul').children().at(4).simulate('click');
    wrapper.find('ul').children().at(6).simulate('click');  
    
    const gameOverText = wrapper.find('h3').text().trim();
    expect(gameOverText).toBe('x player is the winner');
  });
})

