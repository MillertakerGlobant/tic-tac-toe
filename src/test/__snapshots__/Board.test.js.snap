// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board />  Component test renders empty Board component 1`] = `
<div
  className="gameWrapper"
>
  Game Board
  <ul
    className="BoardWrapper"
  >
    <BoardItem
      index={0}
      onClick={[Function]}
    />
    <BoardItem
      index={1}
      onClick={[Function]}
    />
    <BoardItem
      index={2}
      onClick={[Function]}
    />
    <BoardItem
      index={3}
      onClick={[Function]}
    />
    <BoardItem
      index={4}
      onClick={[Function]}
    />
    <BoardItem
      index={5}
      onClick={[Function]}
    />
    <BoardItem
      index={6}
      onClick={[Function]}
    />
    <BoardItem
      index={7}
      onClick={[Function]}
    />
    <BoardItem
      index={8}
      onClick={[Function]}
    />
  </ul>
</div>
`;
